common/models/user.js

```diff
@@ -468,130 +468,143 @@ module.exports = function(User) {
+      verifyOptions.mailer || this.constructor.email || registry.getModelByType(loopback.Email);
 
     // Set a default token generation function if one is not provided
-    var tokenGenerator = options.generateVerificationToken || User.generateVerificationToken;
+    var tokenGenerator = verifyOptions.generateVerificationToken || User.generateVerificationToken;
     assert(typeof tokenGenerator === 'function', 'generateVerificationToken must be a function');
 
-    tokenGenerator(user, function(err, token) {
-      if (err) { return fn(err); }
+    // argument "options" is passed depending on tokenGenerator function requirements
+    if (tokenGenerator.length == 3) {
+      tokenGenerator(user, options, addTokenToUserAndSave);
+    } else {
+      tokenGenerator(user, addTokenToUserAndSave);
+    }
 
+    function addTokenToUserAndSave(err, token) {
+      if (err) return cb(err);
       user.verificationToken = token;
-      user.save(function(err) {
-        if (err) {
-          fn(err);
-        } else {
-          sendEmail(user);
-        }
+      user.save(options, function(err) {
+        if (err) return cb(err);
+        sendEmail(user);
       });
-    });
+    }
 
     // TODO - support more verification types
     function sendEmail(user) {
-      options.verifyHref += '&token=' + user.verificationToken;
+      verifyOptions.verifyHref += '&token=' + user.verificationToken;
 
-      options.verificationToken = user.verificationToken;
+      verifyOptions.verificationToken = user.verificationToken;
 
-      options.text = options.text || g.f('Please verify your email by opening ' +
-        'this link in a web browser:\n\t%s', options.verifyHref);
+      verifyOptions.text = verifyOptions.text || g.f('Please verify your email by opening ' +
+        'this link in a web browser:\n\t%s', verifyOptions.verifyHref);
 
-      options.text = options.text.replace(/\{href\}/g, options.verifyHref);
+      verifyOptions.text = verifyOptions.text.replace(/\{href\}/g, verifyOptions.verifyHref);
 
-      options.to = options.to || user.email;
+      verifyOptions.to = verifyOptions.to || user.email;
 
-      options.subject = options.subject || g.f('Thanks for Registering');
+      verifyOptions.subject = verifyOptions.subject || g.f('Thanks for Registering');
 
-      options.headers = options.headers || {};
+      verifyOptions.headers = verifyOptions.headers || {};
 
-      options.templateFn(options, function(err, html) {
-        if (err) {
-          fn(err);
-        } else {
-          setHtmlContentAndSend(html);
-        }
-      });
+      // argument "options" is passed depending on templateFn function requirements
+      if (verifyOptions.templateFn.length == 3) {
+        verifyOptions.templateFn(verifyOptions, options, setHtmlContentAndSend);
+      } else {
+        verifyOptions.templateFn(verifyOptions, setHtmlContentAndSend);
+      }
+
+      function setHtmlContentAndSend(err, html) {
+        if (err) return cb(err);
```

Comment:
adding the handling of templateFn callback error inside this function to ease the call to templateFn with 2 or 3 args